tools/serverpod_cli/lib/src/analyzer/models/validation/restrictions.dart

```diff
@@ -731,35 +731,38 @@ class Restrictions {
     dynamic content,
     SourceSpan? span,
   ) {
-    var errors = <SourceSpanSeverityException>[];
-    var definition = documentDefinition;
-
-    if (definition is! ClassDefinition) return errors;
+    var classDefinition = documentDefinition;
+    if (classDefinition is! ClassDefinition) return const [];
 
-    var field = definition.findField(parentNodeName);
-    if (field == null) return errors;
-    var type = field.type.className;
+    var field = classDefinition.findField(parentNodeName);
+    if (field == null) return const [];
 
-    if (AnalyzeChecker.isIdType(type) &&
-        !AnalyzeChecker.isParentDefined(content)) {
-      errors.add(SourceSpanSeverityException(
-        'The "parent" property must be defined on id fields.',
-        span,
-      ));
+    var type = field.type;
+    if (type.isIdType && !AnalyzeChecker.isParentDefined(content)) {
+      return [
+        SourceSpanSeverityException(
+          'The "parent" property must be defined on id fields.',
+          span,
+        )
+      ];
     }
 
-    if (!AnalyzeChecker.isFieldDefined(content)) {
-      var isOptional = AnalyzeChecker.isOptionalDefined(content);
-      var isServerOnly = field.scope == ModelFieldScopeDefinition.serverOnly;
-      if (isServerOnly && !isOptional) {
-        errors.add(SourceSpanSeverityException(
+    var relation = field.relation;
+    if (relation is! ObjectRelationDefinition) return const [];
```

Comment:
This is core of the fix. Id relations are explicit integer fields, and the correctness wrt. `serverOnly` is already handled. We only need to consider object relations.